File: src/core/acp/docker/process-manager.ts
import { getServerBridge } from "@/core/platform";
import type { NotificationHandler } from "@/core/acp/processer";
import os from "os";
import path from "path";
import fsSync from "fs";
import {
  DEFAULT_DOCKER_AGENT_IMAGE,
  findAvailablePort,
  generateContainerName,
  sanitizeEnvForLogging,
  shellEscape,
} from "./utils";
import type { DockerContainerConfig, DockerContainerInfo } from "./types";

// Temporary directory for storing auth.json files
const DOCKER_OPENCODE_TMP_DIR = path.join(os.tmpdir(), "routa-opencode-auth");

const DEFAULT_CONTAINER_PORT = 4321;
const DEFAULT_HEALTH_TIMEOUT_MS = 30_000;
const CONTAINER_IDLE_TIMEOUT_MS = 5 * 60 * 1000; // 5 minutes

interface PersistentContainerInfo extends DockerContainerInfo {
  lastUsedAt: Date;
  sessionCount: number;
}

export class DockerProcessManager {
  private static instance: DockerProcessManager | null = null;
  private containers = new Map<string, DockerContainerInfo>();
  private usedPorts = new Set<number>();

  // Container reuse support
  private persistentContainer: PersistentContainerInfo | null = null;
  private idleTimeoutHandle: NodeJS.Timeout | null = null;

  static getInstance(): DockerProcessManager {
    if (!DockerProcessManager.instance) {
      DockerProcessManager.instance = new DockerProcessManager();
    }
    return DockerProcessManager.instance;
  }

  listContainers(): DockerContainerInfo[] {
    return Array.from(this.containers.values());
  }

  getContainer(sessionId: string): DockerContainerInfo | undefined {
    return this.containers.get(sessionId);
  }

  /**
   * Acquire a container for a session, reusing the persistent container if healthy.
   * This is the main entry point for container lifecycle management with reuse support.
   */
  async acquireContainer(config: DockerContainerConfig): Promise<DockerContainerInfo> {
    // Try to reuse persistent container if available and healthy
    if (this.persistentContainer && await this.isContainerHealthy(this.persistentContainer)) {
      console.log(
        `[DockerProcessManager] Reusing persistent container ${this.persistentContainer.containerName} ` +
        `for session ${config.sessionId} (sessions: ${this.persistentContainer.sessionCount + 1})`
      );

      // Update session tracking
      this.persistentContainer.lastUsedAt = new Date();
      this.persistentContainer.sessionCount++;

      // Map this session to the persistent container
      const sessionInfo: DockerContainerInfo = {
        ...this.persistentContainer,
        sessionId: config.sessionId,
      };
      this.containers.set(config.sessionId, sessionInfo);

      // Cancel idle timeout since container is now in use
      this.cancelIdleTimeout();

      return sessionInfo;
    }

    // No healthy persistent container available, start a new one
    return this.startContainer(config);
  }

  /**
   * Check if a container is healthy by querying its health endpoint.
   */
  private async isContainerHealthy(info: DockerContainerInfo): Promise<boolean> {
    try {
      const healthUrl = `http://127.0.0.1:${info.hostPort}/health`;
      const res = await fetch(healthUrl, {
        cache: "no-store",
        signal: AbortSignal.timeout(3000), // 3 second timeout
      });
      return res.ok;
    } catch {
      return false;
    }
  }

  /**
   * Start a new idle timeout for the persistent container.
   */
  private scheduleIdleTimeout(): void {
    this.cancelIdleTimeout();

    this.idleTimeoutHandle = setTimeout(async () => {
      if (this.persistentContainer) {
        const idleTime = Date.now() - this.persistentContainer.lastUsedAt.getTime();
        console.log(
          `[DockerProcessManager] Persistent container ${this.persistentContainer.containerName} ` +
          `idle for ${Math.floor(idleTime / 1000)}s, stopping...`
        );
        await this.stopPersistentContainer();
      }
    }, CONTAINER_IDLE_TIMEOUT_MS);
  }

  /**
   * Cancel the idle timeout if it exists.
   */
  private cancelIdleTimeout(): void {
    if (this.idleTimeoutHandle) {
      clearTimeout(this.idleTimeoutHandle);
      this.idleTimeoutHandle = null;
    }
  }

  /**
   * Stop the persistent container and clean up resources.
   */
  private async stopPersistentContainer(): Promise<void> {
    if (!this.persistentContainer) return;

    const bridge = getServerBridge();
    const containerName = this.persistentContainer.containerName;

    try {
      await bridge.process.exec(`docker stop -t 10 ${shellEscape(containerName)}`, { timeout: 15_000 });
    } catch {
      try {
        await bridge.process.exec(`docker kill ${shellEscape(containerName)}`, { timeout: 8_000 });
      } catch {
        // Ignore kill failure
      }
    }

    try {
      await bridge.process.exec(`docker rm -f ${shellEscape(containerName)}`, { timeout: 8_000 });
    } catch {
      // Ignore cleanup failure
    }

    this.usedPorts.delete(this.persistentContainer.hostPort);
    this.persistentContainer = null;
    this.cancelIdleTimeout();

    console.log(`[DockerProcessManager] Stopped persistent container ${containerName}`);
  }

  async startContainer(config: DockerContainerConfig): Promise<DockerContainerInfo> {
    const bridge = getServerBridge();
    const containerPort = config.containerPort ?? DEFAULT_CONTAINER_PORT;
    const hostPort = await findAvailablePort(this.usedPorts);
    const containerName = generateContainerName(config.sessionId);

    const labels = {
      "routa.managed": "true",
      "routa.session": config.sessionId,
      ...(config.labels ?? {}),
    };

    const envEntries = Object.entries(config.env ?? {}).filter(([, value]) => typeof value === "string") as Array<[string, string]>;
    const sanitizedEnv = sanitizeEnvForLogging(config.env);

    const runParts: string[] = [
      "docker run -d",
      "--rm",
      `--name ${shellEscape(containerName)}`,
      `-p ${hostPort}:${containerPort}`,
      // Resource limits to prevent runaway processes
      "--memory", "2g",
      "--cpus", "2",
      "--pids-limit", "100",
      // Allow the container to reach the host machine (e.g. for MCP at localhost:PORT).
      // On Docker Desktop (Mac/Windows) this resolves automatically; on Linux we
      // need to explicitly map the host-gateway alias.
      "--add-host=host.docker.internal:host-gateway",
      "-w /workspace",
      `-v ${shellEscape(`${config.workspacePath}:/workspace`)}`,
    ];

    // Mount host SSH keys so git SSH operations (clone, push, pull) work inside
    // the container. Mounted read-only to avoid accidental modification.
    const sshDir = path.join(os.homedir(), ".ssh");
    if (fsSync.existsSync(sshDir)) {
      runParts.push(`-v ${shellEscape(`${sshDir}:/root/.ssh:ro`)}`);
    }

    // Mount host git config so commits carry the correct author identity and
    // credential helpers are available for HTTPS operations.
    const gitConfigFile = path.join(os.homedir(), ".gitconfig");
    if (fsSync.existsSync(gitConfigFile)) {
      runParts.push(`-v ${shellEscape(`${gitConfigFile}:/root/.gitconfig:ro`)}`);
    }

    // Derive the Routa MCP URL that is reachable from inside the container.
    // host.docker.internal resolves to the host machine; PORT is the Next.js port.
    const routaPort = process.env.PORT || "3000";
    const routaMcpUrl = `http://host.docker.internal:${routaPort}/api/mcp`;
    runParts.push(`-e ${shellEscape(`ROUTA_MCP_URL=${routaMcpUrl}`)}`);

    // Forward GitHub token for HTTPS git authentication when available.
    if (process.env.GITHUB_TOKEN) {
      runParts.push(`-e GITHUB_TOKEN=${shellEscape(process.env.GITHUB_TOKEN)}`);
    }

    // Forward common AI provider API keys from the host so that the opencode
    // agent can authenticate against Anthropic / OpenAI / etc. without the user
    // having to re-configure credentials inside the container.
    const providerKeyVars = [
      "ANTHROPIC_API_KEY",
      "ANTHROPIC_AUTH_TOKEN",
      "OPENAI_API_KEY",
      "OPENAI_API_BASE",
      "OPENAI_BASE_URL",
      "GEMINI_API_KEY",
      "OPENROUTER_API_KEY",
      "XAI_API_KEY",
      "AZURE_OPENAI_API_KEY",
      "AZURE_OPENAI_ENDPOINT",
    ];
    for (const key of providerKeyVars) {
      if (process.env[key]) {
        runParts.push(`-e ${shellEscape(`${key}=${process.env[key]}`)}`);
      }
    }

    for (const [key, value] of Object.entries(labels)) {
      runParts.push(`--label ${shellEscape(`${key}=${value}`)}`);
    }

    for (const volume of config.additionalVolumes ?? []) {
      runParts.push(`-v ${shellEscape(`${volume.hostPath}:${volume.containerPath}`)}`);
    }

    for (const [key, value] of envEntries) {
      runParts.push(`-e ${shellEscape(`${key}=${value}`)}`);
    }

    // Mount auth.json if provided — write to temp file and bind-mount into container
    let authJsonTempFile: string | null = null;
    if (config.authJson?.trim()) {
      try {
        fsSync.mkdirSync(DOCKER_OPENCODE_TMP_DIR, { recursive: true });
        authJsonTempFile = path.join(DOCKER_OPENCODE_TMP_DIR, `auth-${config.sessionId}.json`);
        fsSync.writeFileSync(authJsonTempFile, config.authJson, "utf-8");
        // Opencode reads from ~/.local/share/opencode/auth.json
        runParts.push(`-v ${shellEscape(`${authJsonTempFile}:/root/.local/share/opencode/auth.json:ro`)}`);
        console.log(`[DockerProcessManager] Mounted auth.json from ${authJsonTempFile}`);
      } catch (err) {
        console.error(`[DockerProcessManager] Failed to write auth.json: ${err instanceof Error ? err.message : String(err)}`);
      }
    }

    runParts.push(shellEscape(config.image || DEFAULT_DOCKER_AGENT_IMAGE));

    const runCommand = runParts.join(" ");

    try {
      const { stdout } = await bridge.process.exec(runCommand, { timeout: 30_000 });
      const containerId = stdout.trim();

      const info: DockerContainerInfo = {
        sessionId: config.sessionId,
        containerId,
        containerName,
        hostPort,
        containerPort,
        image: config.image || DEFAULT_DOCKER_AGENT_IMAGE,
        workspacePath: config.workspacePath,
        createdAt: new Date(),
      };

      this.containers.set(config.sessionId, info);
      this.usedPorts.add(hostPort);

      // Set as persistent container for reuse
      this.persistentContainer = {
        ...info,
        lastUsedAt: new Date(),
        sessionCount: 1,
      };

      console.log(
        `[DockerProcessManager] Started container ${containerName} on port ${hostPort} ` +
        `(image: ${info.image}, env: ${JSON.stringify(sanitizedEnv)}, reusable: true)`
      );

      return info;
    } catch (err) {
      throw new Error(`Failed to start Docker container: ${err instanceof Error ? err.message : String(err)}`);
    }
  }

  async waitForHealthy(
    sessionId: string,
    timeoutMs = DEFAULT_HEALTH_TIMEOUT_MS,
    onProgress?: NotificationHandler,
  ): Promise<void> {
    const info = this.containers.get(sessionId);
    if (!info) {
      throw new Error(`No managed Docker container for session ${sessionId}`);
    }

    const emitLog = (text: string) => {
      if (!onProgress) return;
      onProgress({
        jsonrpc: "2.0",
        method: "session/update",
        params: {
          sessionId,
          update: {
            sessionUpdate: "process_output",
            source: "docker",
            data: text,
            displayName: "Docker",
          },
        },
      });
    };

    const start = Date.now();
    const healthUrl = `http://127.0.0.1:${info.hostPort}/health`;
    let lastTickSec = -1;

    emitLog(`Starting container ${info.containerName} on port ${info.hostPort}...\r\n`);

    while (Date.now() - start < timeoutMs) {
      try {
        const res = await fetch(healthUrl, { cache: "no-store" });
        if (res.ok) {
          emitLog(`Container is healthy ✓\r\n`);
          return;
        }
      } catch {
        // Ignore transient startup failures while container is booting.
      }

      const elapsed = Math.floor((Date.now() - start) / 1000);
      if (elapsed !== lastTickSec && elapsed % 3 === 0) {
        lastTickSec = elapsed;
        // Stream recent container logs every 3 seconds
        try {
          const bridge = getServerBridge();
          const { stdout, stderr } = await bridge.process.exec(
            `docker logs --tail 10 ${shellEscape(info.containerName)}`,
            { timeout: 3_000 },
          );
          const combined = (stdout + stderr).trim();
          if (combined) {
            emitLog(combined + "\r\n");
          } else {
            emitLog(`Waiting for container health check... (${elapsed}s)\r\n`);
          }
        } catch {
          emitLog(`Waiting for container health check... (${elapsed}s)\r\n`);
        }
      }

      await new Promise((resolve) => setTimeout(resolve, 1_000));
    }

    const logs = await this.getContainerLogs(info.containerName);
    throw new Error(
      `Docker container health check timeout after ${timeoutMs}ms for ${info.containerName}. ` +
      `Health endpoint: ${healthUrl}. Logs:\n${logs}`
    );
  }

  async stopContainer(sessionId: string): Promise<void> {
    const info = this.containers.get(sessionId);
    if (!info) return;

    // Remove session mapping
    this.containers.delete(sessionId);

    // If this is the persistent container, don't stop it immediately
    // Instead, schedule idle timeout for potential reuse
    if (this.persistentContainer && info.containerName === this.persistentContainer.containerName) {
      console.log(
        `[DockerProcessManager] Session ${sessionId} ended, keeping persistent container ` +
        `${info.containerName} alive for reuse (idle timeout: ${CONTAINER_IDLE_TIMEOUT_MS / 1000}s)`
      );
      this.scheduleIdleTimeout();
      return;
    }

    // Non-persistent container, stop it immediately
    const bridge = getServerBridge();

    try {
      await bridge.process.exec(`docker stop -t 10 ${shellEscape(info.containerName)}`, { timeout: 15_000 });
    } catch {
      try {
        await bridge.process.exec(`docker kill ${shellEscape(info.containerName)}`, { timeout: 8_000 });
      } catch {
        // Ignore kill failure; rm -f below is best effort.
      }
    }

    try {
      await bridge.process.exec(`docker rm -f ${shellEscape(info.containerName)}`, { timeout: 8_000 });
    } catch {
      // Ignore cleanup failure for already-removed containers.
    }

    this.usedPorts.delete(info.hostPort);
  }

  async stopAll(): Promise<void> {
    const ids = Array.from(this.containers.keys());
    for (const sessionId of ids) {
      await this.stopContainer(sessionId);
    }

    // Also stop the persistent container
    await this.stopPersistentContainer();

    this.containers.clear();
    this.usedPorts.clear();
  }

  private async getContainerLogs(containerName: string): Promise<string> {
    const bridge = getServerBridge();
    try {
      const { stdout, stderr } = await bridge.process.exec(
        `docker logs --tail 200 ${shellEscape(containerName)}`,
        { timeout: 5_000 },
      );
      return `${stdout}${stderr ? `\n${stderr}` : ""}`.trim();
    } catch (err) {
      return `Failed to read logs: ${err instanceof Error ? err.message : String(err)}`;
    }
  }
}

export function getDockerProcessManager(): DockerProcessManager {
  return DockerProcessManager.getInstance();
}

File: crates/routa-core/src/acp/docker/process_manager.rs
//! Docker process manager for container lifecycle management.
//!
//! Mirrors the TypeScript `DockerProcessManager` in `src/core/acp/docker/process-manager.ts`.

use super::types::{DockerContainerConfig, DockerContainerInfo};
use super::utils::{
    find_available_port, generate_container_name, sanitize_env_for_logging, shell_escape,
    DEFAULT_CONTAINER_PORT, DEFAULT_DOCKER_AGENT_IMAGE, DEFAULT_HEALTH_TIMEOUT_MS,
};
use chrono::{DateTime, Utc};
use std::collections::{HashMap, HashSet};
use std::path::PathBuf;
use std::process::Stdio;
use std::sync::Arc;
use std::time::Duration;
use tokio::process::Command;
use tokio::sync::RwLock;

const CONTAINER_IDLE_TIMEOUT_MS: u64 = 5 * 60 * 1000; // 5 minutes

/// Extended container info for persistent container tracking
#[derive(Debug, Clone)]
struct PersistentContainerInfo {
    info: DockerContainerInfo,
    last_used_at: DateTime<Utc>,
    session_count: usize,
}

/// Docker process manager for container lifecycle management.
pub struct DockerProcessManager {
    containers: Arc<RwLock<HashMap<String, DockerContainerInfo>>>,
    used_ports: Arc<RwLock<HashSet<u16>>>,
    persistent_container: Arc<RwLock<Option<PersistentContainerInfo>>>,
}

impl Default for DockerProcessManager {
    fn default() -> Self {
        Self::new()
    }
}

impl DockerProcessManager {
    /// Create a new DockerProcessManager instance.
    pub fn new() -> Self {
        Self {
            containers: Arc::new(RwLock::new(HashMap::new())),
            used_ports: Arc::new(RwLock::new(HashSet::new())),
            persistent_container: Arc::new(RwLock::new(None)),
        }
    }

    /// List all managed containers.
    pub async fn list_containers(&self) -> Vec<DockerContainerInfo> {
        self.containers.read().await.values().cloned().collect()
    }

    /// Get a container by session ID.
    pub async fn get_container(&self, session_id: &str) -> Option<DockerContainerInfo> {
        self.containers.read().await.get(session_id).cloned()
    }

    /// Acquire a container for a session, reusing the persistent container if healthy.
    /// This is the main entry point for container lifecycle management with reuse support.
    pub async fn acquire_container(
        &self,
        config: DockerContainerConfig,
    ) -> Result<DockerContainerInfo, String> {
        // Try to reuse persistent container if available and healthy
        let should_reuse = {
            let persistent = self.persistent_container.read().await;
            if let Some(ref persistent_info) = *persistent {
                self.is_container_healthy(&persistent_info.info).await
            } else {
                false
            }
        };

        if should_reuse {
            // Update persistent container tracking
            let mut persistent_write = self.persistent_container.write().await;
            if let Some(ref mut p) = *persistent_write {
                tracing::info!(
                    "[DockerProcessManager] Reusing persistent container {} for session {} (sessions: {})",
                    p.info.container_name,
                    config.session_id,
                    p.session_count + 1
                );

                p.last_used_at = Utc::now();
                p.session_count += 1;

                // Map this session to the persistent container
                let session_info = DockerContainerInfo {
                    session_id: config.session_id.clone(),
                    ..p.info.clone()
                };
                self.containers.write().await.insert(config.session_id.clone(), session_info.clone());

                return Ok(session_info);
            }
        }

        // No healthy persistent container available, start a new one
        self.start_container(config).await
    }

    /// Check if a container is healthy by querying its health endpoint.
    async fn is_container_healthy(&self, info: &DockerContainerInfo) -> bool {
        let health_url = format!("http://127.0.0.1:{}/health", info.host_port);

        match tokio::time::timeout(
            Duration::from_secs(3),
            reqwest::get(&health_url)
        ).await {
            Ok(Ok(response)) => response.status().is_success(),
            _ => false,
        }
    }

    /// Start a Docker container for an agent session.
    pub async fn start_container(
        &self,
        config: DockerContainerConfig,
    ) -> Result<DockerContainerInfo, String> {
        let container_port = config.container_port.unwrap_or(DEFAULT_CONTAINER_PORT);
        let used_ports = self.used_ports.read().await.clone();
        let host_port = find_available_port(&used_ports).await?;
        let container_name = generate_container_name(&config.session_id);

        let mut labels = HashMap::new();
        labels.insert("routa.managed".to_string(), "true".to_string());
        labels.insert("routa.session".to_string(), config.session_id.clone());
        if let Some(extra_labels) = &config.labels {
            labels.extend(extra_labels.clone());
        }

        let sanitized_env = sanitize_env_for_logging(config.env.as_ref());

        // Build docker run command
        let mut run_parts = vec![
            "docker".to_string(),
            "run".to_string(),
            "-d".to_string(),
            "--rm".to_string(),
            format!("--name={}", container_name),
            format!("-p={}:{}", host_port, container_port),
            // Resource limits to prevent runaway processes
            "--memory=2g".to_string(),
            "--cpus=2".to_string(),
            "--pids-limit=100".to_string(),
            "--add-host=host.docker.internal:host-gateway".to_string(),
            "-w=/workspace".to_string(),
            format!("-v={}:/workspace", shell_escape(&config.workspace_path)),
        ];

        // Mount SSH keys if available
        if let Some(home) = dirs::home_dir() {
            let ssh_dir = home.join(".ssh");
            if ssh_dir.exists() {
                run_parts.push(format!(
                    "-v={}:/root/.ssh:ro",
                    shell_escape(&ssh_dir.to_string_lossy())
                ));
            }

            let gitconfig = home.join(".gitconfig");
            if gitconfig.exists() {
                run_parts.push(format!(
                    "-v={}:/root/.gitconfig:ro",
                    shell_escape(&gitconfig.to_string_lossy())
                ));
            }
        }

        // Set Routa MCP URL
        let routa_port = std::env::var("PORT").unwrap_or_else(|_| "3000".to_string());
        run_parts.push(format!("-e=ROUTA_MCP_URL=http://host.docker.internal:{}/api/mcp", routa_port));

        // Forward provider API keys
        Self::forward_env_vars(&mut run_parts);

        // Add labels
        for (key, value) in &labels {
            run_parts.push(format!("--label={}={}", key, value));
        }

        // Add additional volumes
        if let Some(volumes) = &config.additional_volumes {
            for vol in volumes {
                run_parts.push(format!(
                    "-v={}:{}",
                    shell_escape(&vol.host_path),
                    shell_escape(&vol.container_path)
                ));
            }
        }

        // Add environment variables
        if let Some(env) = &config.env {
            for (key, value) in env {
                run_parts.push(format!("-e={}={}", key, shell_escape(value)));
            }
        }

        // Mount auth.json if provided
        if let Some(auth_json) = &config.auth_json {
            if !auth_json.trim().is_empty() {
                if let Some(temp_file) = self.write_auth_json(&config.session_id, auth_json).await? {
                    run_parts.push(format!(
                        "-v={}:/root/.local/share/opencode/auth.json:ro",
                        shell_escape(&temp_file.to_string_lossy())
                    ));
                }
            }
        }

        // Add image
        let image = if config.image.is_empty() {
            DEFAULT_DOCKER_AGENT_IMAGE.to_string()
        } else {
            config.image.clone()
        };
        run_parts.push(image.clone());

        // Execute docker run
        let output = self.run_docker_command(&run_parts).await?;
        let container_id = output.trim().to_string();

        let info = DockerContainerInfo {
            session_id: config.session_id.clone(),
            container_id,
            container_name: container_name.clone(),
            host_port,
            container_port,
            image,
            workspace_path: config.workspace_path,
            created_at: Utc::now(),
        };

        self.containers
            .write()
            .await
            .insert(config.session_id.clone(), info.clone());
        self.used_ports.write().await.insert(host_port);

        // Set as persistent container for reuse
        *self.persistent_container.write().await = Some(PersistentContainerInfo {
            info: info.clone(),
            last_used_at: Utc::now(),
            session_count: 1,
        });

        tracing::info!(
            "[DockerProcessManager] Started container {} on port {} (image: {}, env: {:?}, reusable: true)",
            container_name,
            host_port,
            info.image,
            sanitized_env
        );

        Ok(info)
    }

    /// Wait for a container to become healthy.
    pub async fn wait_for_healthy(
        &self,
        session_id: &str,
        timeout_ms: Option<u64>,
    ) -> Result<(), String> {
        let info = self
            .get_container(session_id)
            .await
            .ok_or_else(|| format!("No managed Docker container for session {}", session_id))?;

        let timeout = Duration::from_millis(timeout_ms.unwrap_or(DEFAULT_HEALTH_TIMEOUT_MS));
        let health_url = format!("http://127.0.0.1:{}/health", info.host_port);
        let start = std::time::Instant::now();

        tracing::info!(
            "[DockerProcessManager] Starting container {} on port {}...",
            info.container_name,
            info.host_port
        );

        while start.elapsed() < timeout {
            match reqwest::get(&health_url).await {
                Ok(resp) if resp.status().is_success() => {
                    tracing::info!("[DockerProcessManager] Container is healthy ✓");
                    return Ok(());
                }
                _ => {
                    // Transient failure, retry
                }
            }

            tokio::time::sleep(Duration::from_secs(1)).await;
        }

        let logs = self.get_container_logs(&info.container_name).await;
        Err(format!(
            "Docker container health check timeout after {}ms for {}. Health endpoint: {}. Logs:\n{}",
            timeout.as_millis(),
            info.container_name,
            health_url,
            logs
        ))
    }

    /// Stop and remove a container by session ID.
    pub async fn stop_container(&self, session_id: &str) -> Result<(), String> {
        let info = match self.containers.read().await.get(session_id).cloned() {
            Some(info) => info,
            None => return Ok(()),
        };

        // Remove session mapping
        self.containers.write().await.remove(session_id);

        // If this is the persistent container, don't stop it immediately
        // Instead, schedule idle timeout for potential reuse
        let persistent = self.persistent_container.read().await;
        if let Some(ref persistent_info) = *persistent {
            if info.container_name == persistent_info.info.container_name {
                drop(persistent);
                tracing::info!(
                    "[DockerProcessManager] Session {} ended, keeping persistent container {} alive for reuse (idle timeout: {}s)",
                    session_id,
                    info.container_name,
                    CONTAINER_IDLE_TIMEOUT_MS / 1000
                );
                // Note: In Rust, we don't implement the idle timeout mechanism here
                // as it would require spawning background tasks. The container will
                // be reused by subsequent sessions or stopped when stop_all is called.
                return Ok(());
            }
        }
        drop(persistent);

        // Non-persistent container, stop it immediately
        // Try graceful stop first
        let _ = self
            .run_docker_command(&["docker", "stop", "-t", "10", &info.container_name])
            .await;

        // Force kill if needed
        let _ = self
            .run_docker_command(&["docker", "kill", &info.container_name])
            .await;

        // Force remove
        let _ = self
            .run_docker_command(&["docker", "rm", "-f", &info.container_name])
            .await;

        self.used_ports.write().await.remove(&info.host_port);

        Ok(())
    }

    /// Stop the persistent container and clean up resources.
    async fn stop_persistent_container(&self) -> Result<(), String> {
        let mut persistent = self.persistent_container.write().await;
        if let Some(persistent_info) = persistent.take() {
            let container_name = &persistent_info.info.container_name;

            // Try graceful stop first
            let _ = self
                .run_docker_command(&["docker", "stop", "-t", "10", container_name])
                .await;

            // Force kill if needed
            let _ = self
                .run_docker_command(&["docker", "kill", container_name])
                .await;

            // Force remove
            let _ = self
                .run_docker_command(&["docker", "rm", "-f", container_name])
                .await;

            self.used_ports.write().await.remove(&persistent_info.info.host_port);

            tracing::info!("[DockerProcessManager] Stopped persistent container {}", container_name);
        }
        Ok(())
    }

    /// Stop all managed containers.
    pub async fn stop_all(&self) -> Result<(), String> {
        let session_ids: Vec<String> = self.containers.read().await.keys().cloned().collect();

        for session_id in session_ids {
            let _ = self.stop_container(&session_id).await;
        }

        // Also stop the persistent container
        let _ = self.stop_persistent_container().await;

        self.containers.write().await.clear();
        self.used_ports.write().await.clear();

        Ok(())
    }

    /// Run a docker command and return stdout.
    async fn run_docker_command(&self, args: &[impl AsRef<str>]) -> Result<String, String> {
        let args: Vec<&str> = args.iter().map(|s| s.as_ref()).collect();
        if args.is_empty() {
            return Err("Empty command".to_string());
        }

        let result = tokio::time::timeout(
            Duration::from_secs(30),
            Command::new(&args[0])
                .args(&args[1..])
                .stdout(Stdio::piped())
                .stderr(Stdio::piped())
                .output(),
        )
        .await;

        match result {
            Ok(Ok(output)) if output.status.success() => {
                Ok(String::from_utf8_lossy(&output.stdout).to_string())
            }
            Ok(Ok(output)) => {
                let stderr = String::from_utf8_lossy(&output.stderr);
                Err(format!("Docker command failed: {}", stderr))
            }
            Ok(Err(e)) => Err(format!("Failed to run docker: {}", e)),
            Err(_) => Err("Docker command timed out".to_string()),
        }
    }

    /// Get container logs.
    async fn get_container_logs(&self, container_name: &str) -> String {
        match self
            .run_docker_command(&["docker", "logs", "--tail", "200", container_name])
            .await
        {
            Ok(logs) => logs,
            Err(e) => format!("Failed to read logs: {}", e),
        }
    }

    /// Forward environment variables from host to container.
    fn forward_env_vars(run_parts: &mut Vec<String>) {
        let provider_keys = [
            "GITHUB_TOKEN",
            "ANTHROPIC_API_KEY",
            "ANTHROPIC_AUTH_TOKEN",
            "OPENAI_API_KEY",
            "OPENAI_API_BASE",
            "OPENAI_BASE_URL",
            "GEMINI_API_KEY",
            "OPENROUTER_API_KEY",
            "XAI_API_KEY",
            "AZURE_OPENAI_API_KEY",
            "AZURE_OPENAI_ENDPOINT",
        ];

        for key in provider_keys {
            if let Ok(value) = std::env::var(key) {
                run_parts.push(format!("-e={}={}", key, shell_escape(&value)));
            }
        }
    }

    /// Write auth.json to a temporary file.
    async fn write_auth_json(
        &self,
        session_id: &str,
        auth_json: &str,
    ) -> Result<Option<PathBuf>, String> {
        let temp_dir = std::env::temp_dir().join("routa-opencode-auth");
        tokio::fs::create_dir_all(&temp_dir)
            .await
            .map_err(|e| format!("Failed to create temp directory: {}", e))?;

        let temp_file = temp_dir.join(format!("auth-{}.json", session_id));
        tokio::fs::write(&temp_file, auth_json)
            .await
            .map_err(|e| format!("Failed to write auth.json: {}", e))?;

        tracing::info!(
            "[DockerProcessManager] Mounted auth.json from {:?}",
            temp_file
        );

        Ok(Some(temp_file))
    }
}

